drivers/led/is31fl3194.c

```diff
@@ -0,0 +1,357 @@
+};
+
+static const struct led_info *is31fl3194_led_to_info(const struct is31fl3194_config *config,
+						     uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->led_infos[led];
+	}
+
+	return NULL;
+}
+
+static int is31fl3194_get_info(const struct device *dev,
+			       uint32_t led,
+			       const struct led_info **info_out)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *info = is31fl3194_led_to_info(config, led);
+
+	if (info == NULL) {
+		return -EINVAL;
+	}
+
+	*info_out = info;
+	return 0;
+}
+
+static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t num_colors,
+				const uint8_t *color)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *info = is31fl3194_led_to_info(config, led);
+	int ret;
+
+	if (info == NULL) {
+		return -ENODEV;
+	}
+
+	if (info->num_colors != 3) {
+		return -ENOTSUP;
+	}
+
+	if (num_colors != 3) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < 3; i++) {
+		uint8_t value;
+
+		switch (info->color_mapping[i]) {
+		case LED_COLOR_ID_RED:
+			value = color[0];
+			break;
+		case LED_COLOR_ID_GREEN:
+			value = color[1];
+			break;
+		case LED_COLOR_ID_BLUE:
+			value = color[2];
+			break;
+		default:
+			/* unreachable: mapping already tested in is31fl3194_check_config */
+			continue;
+		}
+
+		ret = i2c_reg_write_byte_dt(&config->bus, led_channels[i], value);
+		if (ret != 0) {
+			break;
+		}
+	}
+
+	if (ret == 0) {
+		ret = i2c_reg_write_byte_dt(&config->bus,
+					    IS31FL3194_UPDATE_REG,
+					    IS31FL3194_UPDATE_VAL);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("%s: LED write failed: %d", dev->name, ret);
+	}
+
+	return ret;
+}
+
+static int is31fl3194_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *info = is31fl3194_led_to_info(config, led);
+	int ret = 0;
+
+	if (info == NULL) {
+		return -ENODEV;
+	}
+
+	if (info->num_colors != 1) {
+		return -ENOTSUP;
+	}
+
+	if (value > 100) {
+		return -EINVAL;
+	}
+
+	/* Rescale 0..100 to 0..255 */
+	value = value * 255 / 100;
+
+	ret = i2c_reg_write_byte_dt(&config->bus, led_channels[led], value);
+	if (ret == 0) {
+		ret = i2c_reg_write_byte_dt(&config->bus,
+					    IS31FL3194_UPDATE_REG,
+					    IS31FL3194_UPDATE_VAL);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("%s: LED write failed", dev->name);
+	}
+
+	return ret;
+}
+
+static inline int is31fl3194_led_on(const struct device *dev, uint32_t led)
+{
+	return is31fl3194_set_brightness(dev, led, 100);
+}
+
+static inline int is31fl3194_led_off(const struct device *dev, uint32_t led)
+{
+	return is31fl3194_set_brightness(dev, led, 0);
+}
+
+/* Counts red, green, blue channels; returns true if color_id is valid */
+static bool is31fl3194_count_colors(const struct device *dev,
+				    uint8_t color_id, int *rgb_counts)
+{
+	switch (color_id) {
+	case LED_COLOR_ID_RED:
+		rgb_counts[0]++;
+		return true;
```

Comment:
At the expense of the `multiple channels map to ...` error message you can:

`return (++rgb_counts[0] == 1);`

Choose the option you prefer.